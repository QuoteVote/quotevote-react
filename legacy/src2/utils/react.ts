import React, { 
  useState, 
  useEffect, 
  useRef, 
  ReactNode, 
  ChangeEvent, 
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
} from 'react';

export type { 
  ReactNode,
  ChangeEvent,
  KeyboardEvent,
  FormEvent,
  MouseEvent,
  Dispatch,
  SetStateAction
};

export { 
  React as default,
  useState,
  useEffect,
  useRef
};

// Add type declarations for common event handlers
export type InputChangeEvent = ChangeEvent<HTMLInputElement>;
export type TextAreaChangeEvent = ChangeEvent<HTMLTextAreaElement>;
export type ButtonClickEvent = MouseEvent<HTMLButtonElement>;
export type SetState<T> = Dispatch<SetStateAction<T>>;

// Add type declarations for common component props
export interface BaseProps {
  className?: string;
  style?: React.CSSProperties;
  children?: ReactNode;
}

export interface InputProps extends BaseProps {
  value: string;
  onChange: (value: string) => void;
  placeholder?: string;
  disabled?: boolean;
} 