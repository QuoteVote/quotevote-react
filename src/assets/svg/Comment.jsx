import React from 'react'

const CommentIcon = React.forwardRef((props, ref) => (
  <svg xmlns="http://www.w3.org/2000/svg" width="28" height="29" fill="none" viewBox="0 0 28 29" ref={ref} {...props}>
    <path fill="#fff" d="M25.491 24.268l.7 3.733-3.325-2.216c-1.166.466-2.45.7-3.85.7-4.142 0-7.642-2.275-8.692-5.367-.175-.584-.292-1.225-.292-1.867 0-4.025 4.026-7.292 8.984-7.292 2.334 0 4.434.7 6.009 1.867C26.833 15.167 28 17.093 28 19.25c0 1.925-.933 3.675-2.509 5.017z"/>
    <path fill="#CDD6E0" d="M24.387 15.984H13.595v1.459h10.792v-1.459zM26.547 18.727H11.555v1.458h14.992v-1.458zM24.387 21.643H13.595V23.1h10.792v-1.458z"/>
    <path fill="#F2F2F2" d="M12.426 0C5.542 0 0 4.55 0 10.209c0 2.217.875 4.258 2.333 5.892l-1.225 5.483 4.55-2.858c1.926 1.05 4.26 1.633 6.768 1.633 6.883 0 12.425-4.55 12.425-10.209C24.91 4.55 19.31 0 12.426 0z"/>
    <path fill="#84DBFF" d="M5.837 12.25c1.128 0 2.042-.915 2.042-2.042 0-1.128-.914-2.042-2.042-2.042-1.127 0-2.041.914-2.041 2.042 0 1.127.914 2.041 2.041 2.041zM12.428 12.25c1.128 0 2.042-.915 2.042-2.042 0-1.128-.914-2.042-2.042-2.042-1.127 0-2.041.914-2.041 2.042 0 1.127.914 2.041 2.041 2.041zM19.08 12.25c1.128 0 2.042-.915 2.042-2.042 0-1.128-.914-2.042-2.042-2.042-1.127 0-2.042.914-2.042 2.042 0 1.127.915 2.041 2.042 2.041z"/>
  </svg>
))

export default CommentIcon 