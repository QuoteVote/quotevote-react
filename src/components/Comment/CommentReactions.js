import React, { useState } from 'react'
import PropTypes from 'prop-types'
import {
  IconButton, Popover,
} from '@material-ui/core'
import { InsertEmoticon } from '@material-ui/icons'
import { makeStyles } from '@material-ui/core/styles'
import { useSelector } from 'react-redux'
import { useMutation } from '@apollo/react-hooks'
import { Picker } from 'emoji-mart'
import Emoji from 'a11y-react-emoji'
import _ from 'lodash'
import 'emoji-mart/css/emoji-mart.css'
import { ADD_ACTION_REACTION, UPDATE_ACTION_REACTION } from '../../graphql/mutations'
import { GET_ACTION_REACTIONS } from '../../graphql/query'

const useStyles = makeStyles(() => ({
  container: {
    display: 'flex',
    alignItems: 'center',
  },
  emoji: {
    padding: 3,
    display: 'flex',
  },
  reactions: {
    borderRadius: 8,
    backgroundColor: '#F6F1F1',
    padding: '2px 6px',
    marginLeft: 4,
  },
}))

function CommentReactions(props) {
  const userId = useSelector((state) => state.user.data._id)
  const classes = useStyles()
  const [open, setOpen] = useState(false)
  const [anchorEl, setAnchorEl] = useState(null)
  const { actionId, reactions } = props
  const [addReaction] = useMutation(ADD_ACTION_REACTION, {
    onError: (err) => {
      // eslint-disable-next-line no-console
      console.log(err)
    },
    refetchQueries: [{
      query: GET_ACTION_REACTIONS,
      variables: {
        actionId,
      },
    }],
  })

  const [updateReaction] = useMutation(UPDATE_ACTION_REACTION, {
    onError: (err) => {
      // eslint-disable-next-line no-console
      console.log(err)
    },
    refetchQueries: [{
      query: GET_ACTION_REACTIONS,
      variables: {
        actionId,
      },
    }],
  })

  const groupedReactions = _.groupBy(reactions, 'emoji')

  const userReaction = _.find(reactions, { userId }) || null

  // Handle emoji button interaction
  function handleClick(event) {
    setAnchorEl(event.target)
    setOpen(true)
  }

  async function handleEmojiSelect(emoji) {
    const newEmoji = emoji.native
    const reaction = {
      userId,
      actionId,
      emoji: newEmoji,
    }

    if (userReaction !== null) {
      await updateReaction({
        variables: { _id: userReaction._id, emoji: reaction.emoji },
      })
    } else {
      await addReaction({
        variables: { reaction },
      })
    }

    setOpen(false)
  }

  const emojiElements = []

  Object.keys(groupedReactions).map((emoji, _id) => emojiElements.push(
    <div className={classes.reactions} key={_id}>
      <Emoji symbol={emoji} />
      <span>{groupedReactions[emoji].length}</span>
    </div>
  ))

  return (
    <div className={classes.container}>
      <div className={classes.emoji}>
        {emojiElements}
      </div>
      <IconButton onClick={(event) => { handleClick(event) }}>
        <InsertEmoticon />
      </IconButton>
      <Popover
        open={open}
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'top',
          horizontal: 'center',
        }}
        transformOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        onClose={() => setOpen(false)}
      >
        <div className="reactions">
          <Picker showPreview={false} showSkinTones={false} onSelect={handleEmojiSelect} />
        </div>
      </Popover>
    </div>
  )
}

CommentReactions.propTypes = {
  actionId: PropTypes.string,
  reactions: PropTypes.array,
}

export default CommentReactions
