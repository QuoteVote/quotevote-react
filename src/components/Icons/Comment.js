import React from 'react'
import SvgIcon from '@material-ui/core/SvgIcon'

export default function Comment(props) {
  return (
    <SvgIcon xmlns="http://www.w3.org/2000/svg" {...props}>
      <g
        transform="translate(0.000000,598.000000) scale(0.100000,-0.100000)"
        fill="#000000"
        stroke="none"
      >
        <path
          d="M304 5721 c-62 -28 -119 -84 -145 -144 l-24 -52 0 -2545 0 -2545 24
        -53 c29 -64 95 -125 158 -146 63 -22 168 -20 218 3 23 10 397 295 832 634
        l792 617 1733 2 1733 3 46 21 c63 29 110 74 139 131 l25 48 3 1915 2 1915 -31
        63 c-32 66 -85 114 -152 139 -24 9 -677 12 -2672 12 -2562 1 -2641 0 -2681
        -18z m4976 -2106 l0 -1555 -1648 -2 -1648 -3 -47 -25 c-26 -14 -309 -227 -629
        -475 -319 -247 -585 -451 -590 -453 -4 -2 -8 912 -8 2032 l0 2036 2285 0 2285
        0 0 -1555z"
        />
        <path
          d="M1627 4459 c-89 -21 -160 -80 -198 -163 -29 -64 -26 -178 7 -244 29
        -59 79 -106 143 -136 l46 -21 1365 0 c1274 0 1368 1 1403 17 62 28 111 73 143
        130 26 47 29 63 29 138 0 75 -3 91 -29 138 -32 57 -81 102 -143 130 -35 16
        -130 17 -1383 19 -764 0 -1361 -3 -1383 -8z"
        />
        <path
          d="M1599 3325 c-61 -19 -136 -87 -165 -150 -31 -66 -33 -170 -3 -235 27
        -60 83 -117 144 -146 l50 -24 1365 0 1365 0 50 24 c148 69 208 251 129 396
        -31 57 -65 89 -128 123 l-41 22 -1360 2 c-1152 1 -1367 0 -1406 -12z"
        />
      </g>
    </SvgIcon>
  )
}
