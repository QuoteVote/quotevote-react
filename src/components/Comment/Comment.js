import React, { useState } from 'react'
import {
  Card, CardActions, CardContent, CardHeader, IconButton,
} from '@material-ui/core'
import { InsertEmoticon, InsertLink } from '@material-ui/icons'
import { makeStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import copy from 'clipboard-copy'
import SweetAlert from 'react-bootstrap-sweetalert'
import AvatarDisplay from '../Avatar'
import { parseCommentDate } from '../../utils/momentUtils'
import { SET_FOCUSED_COMMENT } from '../../store/ui'
import buttonStyle from '../../assets/jss/material-dashboard-pro-react/components/buttonStyle'

const useStyles = makeStyles((theme) => ({
  content: {
    marginLeft: 60,
    marginRight: 40,
    marginTop: -20,
    marginBottom: -20,
  },
  expand: {
    marginLeft: 'auto',
  },
  created: {
    verticalAlign: 'middle',
    marginTop: 20,
    marginRight: 10,
  },
  root: {
    backgroundColor: theme.palette.background.paper,
    width: '100%',
  },
  selectedRoot: {
    backgroundColor: '#f1e8c1',
    width: '100%',
  },
  ...buttonStyle,
}))

function Comment({ comment, postUrl, selected }) {
  const {
    user, content, created, _id,
  } = comment
  const { username, avatar } = user
  const classes = useStyles()
  const history = useHistory()
  const parsedDate = parseCommentDate(created)
  const dispatch = useDispatch()
  const [open, setOpen] = useState(false)
  const hideAlert = () => {
    setOpen(false)
  }
  const baseUrl = window.location.origin
  const handleCopy = async () => {
    await copy(`${baseUrl}${postUrl}/comment#${_id}`)
    setOpen(true)
  }

  return (
    <Card
      onMouseEnter={() => dispatch(SET_FOCUSED_COMMENT(comment))}
      onMouseLeave={() => dispatch(SET_FOCUSED_COMMENT(null))}
      className={selected ? classes.selectedRoot : classes.root}
    >
      <CardHeader
        avatar={(
          <IconButton
            size="small"
            onClick={() => history.push(`/hhsb/Profile/${username}`)}
          >
            <AvatarDisplay height={40} width={40} {...avatar} />
          </IconButton>
        )}
        subheader={`@${username}`}
        action={(
          <div className={classes.created}>
            <span>{parsedDate}</span>
          </div>
        )}
      />
      <CardContent className={classes.content}>
        <p>{content}</p>
      </CardContent>
      <CardActions disableSpacing>
        <IconButton className={classes.expand}>
          <InsertEmoticon />
        </IconButton>
        <IconButton onClick={handleCopy}>
          <InsertLink />
        </IconButton>
      </CardActions>
      {open && (
        <SweetAlert
          confirmBtnCssClass={`${classes.button} ${classes.success}`}
          success
          onConfirm={hideAlert}
          onCancel={hideAlert}
          title="Comment URL copied!"
          timeout={1000}
        />
      )}
    </Card>
  )
}

Comment.propTypes = {
  comment: PropTypes.object.isRequired,
  postUrl: PropTypes.string,
  selected: PropTypes.bool,
}

export default Comment
