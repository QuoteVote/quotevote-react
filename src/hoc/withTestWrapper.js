import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <Router history={hist}>
            <Component {...props} />
          </Router>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
