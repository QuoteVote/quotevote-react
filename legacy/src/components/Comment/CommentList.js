import React from 'react'
import {
  Card, CardContent, Grid, IconButton, Typography,
} from '@material-ui/core'
import PropTypes from 'prop-types'
import moment from 'moment'
import { Skeleton } from '@material-ui/lab'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import { useLocation } from 'react-router-dom'
import { Filter as FilterIcon } from '../Icons'
import Comment from './Comment'

function CommentList({ comments, loading, postUrl }) {
  const location = useLocation()
  const { hash } = location
  React.useEffect(() => {
    if (!loading && comments.length && hash) {
      const element = document.getElementById(hash)
      element.scrollIntoView({ behavior: 'smooth' })
    }
  }, [hash, loading, comments])
  return (
    <>
      <Grid
        container
        item
        direction="row"
        justify="space-between"
        alignItems="flex-start"
      >
        <Grid item>
          <Typography variant="h6">
            Comments
          </Typography>
        </Grid>
        <Grid item>
          <IconButton style={{ marginLeft: 'auto' }}>
            <FilterIcon
              width="32"
              height="32"
              viewBox="0 0 32 32"
            />
          </IconButton>
        </Grid>
      </Grid>
      {loading && (
        <>
          <Skeleton variant="rect" height={118} />
          <br />
          <Skeleton variant="rect" height={118} />
          <br />
          <Skeleton variant="rect" height={118} />
        </>
      )}
      {comments ? (
        <List
          style={{
            height: comments.length > 2 ? '75vh' : 'auto',
            marginTop: 5,
            width: '100%',
            position: 'relative',
            overflow: 'auto',
          }}
        >
          {comments.sort((a, b) => moment(b.created).diff(moment(a.created))).map((comment) => (
            <ListItem
              id={`#${comment._id}`}
              key={comment._id}
              style={{ height: 'auto' }}
              autoFocus={`#${comment._id}` === hash}
            >
              <Comment
                comment={comment}
                postUrl={postUrl}
                selected={`#${comment._id}` === hash}
              />
            </ListItem>
          ))}
        </List>
      ) : <Card><CardContent>Start the discussion... </CardContent></Card>}

    </>
  )
}

CommentList.propTypes = {
  comments: PropTypes.array.isRequired,
  loading: PropTypes.bool,
  postUrl: PropTypes.string,
}

export default CommentList
