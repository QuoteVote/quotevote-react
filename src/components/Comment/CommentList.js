import React from 'react'
import {
  Card, CardContent, Grid, GridList, GridListTile, IconButton, Typography,
} from '@material-ui/core'
import PropTypes from 'prop-types'
import moment from 'moment'
import { Skeleton } from '@material-ui/lab'
import { Filter as FilterIcon } from '../Icons'
import Comment from './Comment'

function CommentList({ comments, loading }) {
  return (
    <>
      <Grid
        container
        item
        direction="row"
        justify="space-between"
        alignItems="flex-start"
      >
        <Grid item>
          <Typography variant="h6">
            Comments
          </Typography>
        </Grid>
        <Grid item>
          <IconButton style={{ marginLeft: 'auto' }}>
            <FilterIcon
              width="32"
              height="32"
              viewBox="0 0 32 32"
            />
          </IconButton>
        </Grid>
      </Grid>
      {loading && (
        <>
          <Skeleton variant="rect" height={118} />
          <br />
          <Skeleton variant="rect" height={118} />
          <br />
          <Skeleton variant="rect" height={118} />
        </>
      )}
      {comments ? (
        <GridList
          spacing={15}
          cols={1}
          cellHeight={180}
          style={{ height: comments.length > 2 ? '80vh' : 'auto', marginTop: 5 }}
        >
          {comments.sort((a, b) => moment(b.created).diff(moment(a.created))).map((comment) => (
            <GridListTile style={{ height: 'auto' }}>
              <Comment comment={comment} />
            </GridListTile>
          ))}
        </GridList>
      ) : <Card><CardContent>Start the discussion... </CardContent></Card>}

    </>
  )
}

CommentList.propTypes = {
  comments: PropTypes.object.isRequired,
  loading: PropTypes.bool,
}

export default CommentList
