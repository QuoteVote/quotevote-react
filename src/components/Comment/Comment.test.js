import React from 'react'
import { render } from '@testing-library/react'

// Component being tested
import Comment from './Comment'
import withTestWrapper from '../../hoc/withTestWrapper'

const CommentPostWrapper = withTestWrapper(Comment)

describe('Comment test -', () => {
  it('renders correctly', () => {
    const { container } = render(
      <CommentPostWrapper />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
