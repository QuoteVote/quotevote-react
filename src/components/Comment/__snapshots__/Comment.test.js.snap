// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment test - renders correctly 1`] = `null`;
